File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        visited_nodes=set()
        curr=headB
        while(curr!=None):
            visited_nodes.add(curr)
            curr=curr.next
        curr=headA
        while(curr!=None):
            if curr in visited_nodes:
                return curr
            curr=curr.next
        return None

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        std::unordered_set<ListNode*>visited_nodes;
         ListNode* curr=headB;
        while(curr!=nullptr)
        {
            visited_nodes.insert(curr);
            curr=curr->next;
        }  
         curr=headA;
        while(curr!=nullptr) 
        {
            if(visited_nodes.find(curr)!=visited_nodes.end())
            {
                return curr;
            }
            curr=curr->next;
        }    
        return nullptr;
    }
};